File: skills/codex-review/lib/findings-loader.ts
export type Finding = {
  body: string;
  confidence_score: number;
  line: number;
  path: string;
  priority: number;
  reasoning: string;
  start_line: number | null;
  suggestion: string | null;
  title: string;
};

export type Findings = {
  changes: ReadonlyArray<string>;
  effort: string | null;
  files: ReadonlyArray<{ description: string; path: string }>;
  findings: ReadonlyArray<Finding>;
  model: string;
  overall_confidence_score: number;
  overall_correctness: "patch is correct" | "patch is incorrect";
  summary: string;
};

export class FindingsValidationError extends Error {
  override name = "FindingsValidationError";
}

const REQUIRED_TOP_LEVEL = [
  "changes",
  "effort",
  "files",
  "findings",
  "model",
  "overall_confidence_score",
  "overall_correctness",
  "summary",
] as const;

const REQUIRED_FINDING_FIELDS = [
  "body",
  "confidence_score",
  "line",
  "path",
  "priority",
  "reasoning",
  "start_line",
  "suggestion",
  "title",
] as const;

export function parseFindings(jsonText: string): Findings {
  let raw: unknown;
  try {
    raw = JSON.parse(jsonText) as unknown;
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    throw new FindingsValidationError(`findings JSON is not valid JSON: ${message}`);
  }
  if (typeof raw !== "object" || raw === null) {
    throw new FindingsValidationError("findings root must be an object");
  }
  const obj = raw as Record<string, unknown>;
  for (const key of REQUIRED_TOP_LEVEL) {
    if (!(key in obj)) throw new FindingsValidationError(`missing top-level field: ${key}`);
  }
  if (!Array.isArray(obj["findings"])) throw new FindingsValidationError("findings must be an array");
  for (const [i, finding] of (obj["findings"] as ReadonlyArray<unknown>).entries()) {
    if (typeof finding !== "object" || finding === null) {
      throw new FindingsValidationError(`findings[${i}] must be an object`);
    }
    const f = finding as Record<string, unknown>;
    for (const key of REQUIRED_FINDING_FIELDS) {
      if (!(key in f)) throw new FindingsValidationError(`findings[${i}] missing field: ${key}`);
    }
    const conf = f["confidence_score"];
    if (typeof conf !== "number" || conf < 0 || conf > 1) {
      throw new FindingsValidationError(`findings[${i}].confidence_score out of range`);
    }
    const priority = f["priority"];
    if (typeof priority !== "number" || priority < 0 || priority > 3) {
      throw new FindingsValidationError(`findings[${i}].priority out of range`);
    }
  }
  const verdict = obj["overall_correctness"];
  if (verdict !== "patch is correct" && verdict !== "patch is incorrect") {
    throw new FindingsValidationError("overall_correctness must be 'patch is correct' or 'patch is incorrect'");
  }
  const overallConf = obj["overall_confidence_score"];
  if (typeof overallConf !== "number" || overallConf < 0 || overallConf > 1) {
    throw new FindingsValidationError("overall_confidence_score out of range");
  }
  return raw as Findings;
}

File: skills/codex-review/scripts/lib/findings_loader.py
"""Validate parsed findings against the runtime shape of review-output-schema.json.

Range checks on confidence_score and priority, enum check on
overall_correctness, and presence checks on every required field.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Literal


@dataclass(frozen=True)
class Finding:
    body: str
    confidence_score: float
    line: int
    path: str
    priority: int
    reasoning: str
    start_line: int | None
    suggestion: str | None
    title: str


@dataclass(frozen=True)
class FileChange:
    description: str
    path: str


@dataclass(frozen=True)
class Findings:
    changes: tuple[str, ...]
    effort: str | None
    files: tuple[FileChange, ...]
    findings: tuple[Finding, ...]
    model: str
    overall_confidence_score: float
    overall_correctness: Literal["patch is correct", "patch is incorrect"]
    summary: str


class FindingsValidationError(Exception):
    """Raised when findings JSON does not match the schema's runtime contract."""


_REQUIRED_TOP = (
    "changes",
    "effort",
    "files",
    "findings",
    "model",
    "overall_confidence_score",
    "overall_correctness",
    "summary",
)
_REQUIRED_FIELDS = (
    "body",
    "confidence_score",
    "line",
    "path",
    "priority",
    "reasoning",
    "start_line",
    "suggestion",
    "title",
)


def parse_findings(json_text: str) -> Findings:
    try:
        raw = json.loads(json_text)
    except json.JSONDecodeError as exc:
        raise FindingsValidationError(f"findings JSON is not valid JSON: {exc}") from exc
    if not isinstance(raw, dict):
        raise FindingsValidationError("findings root must be an object")
    for key in _REQUIRED_TOP:
        if key not in raw:
            raise FindingsValidationError(f"missing top-level field: {key}")
    if not isinstance(raw["findings"], list):
        raise FindingsValidationError("findings must be an array")
    findings_list: list[Finding] = []
    for index, finding in enumerate(raw["findings"]):
        if not isinstance(finding, dict):
            raise FindingsValidationError(f"findings[{index}] must be an object")
        for field_name in _REQUIRED_FIELDS:
            if field_name not in finding:
                raise FindingsValidationError(f"findings[{index}] missing field: {field_name}")
        confidence = finding["confidence_score"]
        if not isinstance(confidence, (int, float)) or not 0 <= confidence <= 1:
            raise FindingsValidationError(f"findings[{index}].confidence_score out of range")
        priority = finding["priority"]
        if not isinstance(priority, int) or not 0 <= priority <= 3:
            raise FindingsValidationError(f"findings[{index}].priority out of range")
        findings_list.append(
            Finding(
                body=finding["body"],
                confidence_score=float(confidence),
                line=int(finding["line"]),
                path=finding["path"],
                priority=int(priority),
                reasoning=finding["reasoning"],
                start_line=finding["start_line"],
                suggestion=finding["suggestion"],
                title=finding["title"],
            )
        )
    verdict = raw["overall_correctness"]
    if verdict not in ("patch is correct", "patch is incorrect"):
        raise FindingsValidationError("overall_correctness must be 'patch is correct' or 'patch is incorrect'")
    overall_conf = raw["overall_confidence_score"]
    if not isinstance(overall_conf, (int, float)) or not 0 <= overall_conf <= 1:
        raise FindingsValidationError("overall_confidence_score out of range")
    files = tuple(FileChange(description=item["description"], path=item["path"]) for item in raw["files"])
    return Findings(
        changes=tuple(raw["changes"]),
        effort=raw["effort"],
        files=files,
        findings=tuple(findings_list),
        model=raw["model"],
        overall_confidence_score=float(overall_conf),
        overall_correctness=verdict,
        summary=raw["summary"],
    )
